### ISSUE TITLE
Remote / cloud based storage

### ISSUE THREAD

**Issue by jetuk** [pywr/pywr#606]
I'm sure I added this issue before, but I can't find it.

It would be good if some or all of the data in a model could be fetched from remote data repositories. For example Amazon's S3.

This should probably been some sort of wrapper that indicates all local url data is actually in a remote bucket... or something.

**Comment by snorfalorpagus** [comment]
Some options:

1. Declare this out of scope for the model - it should be handled by an external process
2. Parameters that read files should be aware of "s3://" urls (and do some local caching?)
3. A root-level section in the JSON document which lists some files to pre-fetch (+ checksums?)

**Comment by rheinheimer** [comment]
I just wanted to share my strong opinion that this should be within the scope of the model. I'm storing files on S3. Definitely some local caching is essential as well. Currently, I'm using my own custom `read_csv` function within my custom root parameter class so that in my parameter classes I have access to `self.read_csv(...)` which is a wrapper around Pandas `read_csv` function that checks a local redis cache first.

For some reason I thought Pywr used pandas to read CSVs, so that I could just pass whatever path in and it would pass through to pandas.

**Comment by jetuk** [comment]
> I thought Pywr used pandas to read CSVs

It does.

@rheinheimer the redis cache is quite an interesting approach. Does this work well? I think a first implementation would be using a non-cached approached. I assume you are using `boto` for accessing s3?

@snorfalorpagus I guess we have an issue with the JSON format where we have multiple references to the same resource. I wonder if your suggestion of a "resources" section on the JSON would actually be useful for defining shared URLs regardless of the use of the cloud or not. One could update the resource reference in one place rather than in multiple.

**Comment by rheinheimer** [comment]
I was using boto for s3. For AWS S3, reading in data is way to slow for development, but could be an option for production runs. For now, I think downloading data from S3 during preprocessing works well. My general flow is: redis or s3 -> csv files (or maybe h5? haven't tried this yet). I.e. check redis for data. If not there, download from s3. 

This is still probably needed, but in the end not critical since we can develop some custom preprocessing to download from s3.

**Comment by jetuk** [comment]
Yeah, this is definitely a nice to have. I do wonder if the pre-fetch / caching is something that should be handled by some sort of intermediate library? Are there no existing solutions in this space?
